File: internal/taskchannel/mux.go
// Package taskchannel is the transport of coop's in-box task tools: a multiplexer script that runs
// in a helper container beside the box and carries every client connection on the channel's unix
// socket over the helper's stdio, and the host-side demultiplexer that serves each connection.
// It knows nothing about tasks or MCP — internal/taskmcp supplies the session handler — so the
// sandbox owner (internal/box) can import it without reaching the task lifecycle.
package taskchannel

import (
	"bufio"
	"context"
	_ "embed"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
)

const (
	// BoxSocketDir is the box-side mountpoint of the channel volume, and BoxSocketPath the
	// socket inside it that every MCP client connects to with `socat STDIO UNIX-CONNECT:`.
	BoxSocketDir  = "/coop/tasks"
	BoxSocketPath = BoxSocketDir + "/mcp.sock"

	// maxMuxFrameBytes bounds one helper frame. The helper caps a raw client line at 4 MiB and
	// JSON-escapes it, so a legitimate frame can be several times larger; anything beyond this is
	// a broken or hostile helper stream and ends the channel rather than the host's memory.
	maxMuxFrameBytes = 32 << 20
)

// MuxScript is the helper-side multiplexer (mux.js): it listens on the channel socket inside the
// VM and carries every client connection over the helper container's stdio. box.Run writes it
// into the run's artifacts dir and mounts it read-only into the HELPER — never into the box.
//
//go:embed mux.js
var MuxScript string

// SessionHandler serves one client connection over rw until the client hangs up; taskmcp's
// Server.Serve is the one the box uses.
type SessionHandler func(ctx context.Context, rw io.ReadWriter) error

// muxFrame is one line of the helper protocol (see mux.js for the grammar).
type muxFrame struct {
	Conn  int     `json:"c,omitempty"`
	Event string  `json:"e,omitempty"`
	Data  *string `json:"d,omitempty"`
}

// ServeMux demultiplexes the helper stream — frames arriving on from, frames sent on to — into one
// session per client connection, each served by handle. ready is called once when the helper
// reports its socket bound. It returns when from ends (the helper exited), on a protocol
// violation, or when ctx is done; every open session is closed on the way out.
func ServeMux(ctx context.Context, from io.Reader, to io.Writer, ready func(), handle SessionHandler) error {
	m := &mux{handle: handle, to: to, sessions: map[int]*muxSession{}, ready: ready}
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
	defer m.closeAll()
	reader := bufio.NewReaderSize(from, 64<<10)
	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		line, err := readMuxFrame(reader)
		if len(line) > 0 {
			var frame muxFrame
			if jsonErr := json.Unmarshal(line, &frame); jsonErr != nil {
				return fmt.Errorf("task channel: malformed helper frame: %w", jsonErr)
			}
			m.dispatch(ctx, frame)
		}
		if err != nil {
			if errors.Is(err, io.EOF) {
				return nil
			}
			return err
		}
	}
}

func readMuxFrame(reader *bufio.Reader) ([]byte, error) {
	var frame []byte
	for {
		chunk, err := reader.ReadSlice('\n')
		frame = append(frame, chunk...)
		if len(frame) > maxMuxFrameBytes {
			return nil, errors.New("task channel: helper frame exceeds its bound")
		}
		if errors.Is(err, bufio.ErrBufferFull) {
			continue
		}
		if len(frame) > 0 && frame[len(frame)-1] == '\n' {
			frame = frame[:len(frame)-1]
		}
		return frame, err
	}
}

type mux struct {
	handle    SessionHandler
	to        io.Writer
	ready     func()
	readyOnce sync.Once
	writeMu   sync.Mutex
	mu        sync.Mutex
	sessions  map[int]*muxSession
	closing   bool
	wg        sync.WaitGroup
}

// muxSession is one client connection: the server reads what the helper relays through in, and
// writes replies back as frames tagged with the connection id.
type muxSession struct {
	id  int
	in  *io.PipeWriter
	out *io.PipeReader
}

func (m *mux) send(frame muxFrame) error {
	data, err := json.Marshal(frame)
	if err != nil {
		return err
	}
	m.writeMu.Lock()
	defer m.writeMu.Unlock()
	_, err = m.to.Write(append(data, '\n'))
	return err
}

func (m *mux) dispatch(ctx context.Context, frame muxFrame) {
	switch {
	case frame.Event == "ready" && frame.Conn == 0:
		if m.ready != nil {
			m.readyOnce.Do(m.ready)
		}
	case frame.Event == "open":
		m.open(ctx, frame.Conn)
	case frame.Event == "close":
		m.close(frame.Conn)
	case frame.Data != nil:
		m.mu.Lock()
		session := m.sessions[frame.Conn]
		m.mu.Unlock()
		if session != nil {
			// A slow session applies backpressure to the whole helper stream by design: the
			// pipe is unbuffered, and a client that never reads replies should not make the host
			// buffer its requests without bound.
			_, _ = io.WriteString(session.in, *frame.Data+"\n")
		}
	}
}

func (m *mux) open(ctx context.Context, id int) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if _, exists := m.sessions[id]; exists || id <= 0 {
		return
	}
	out, in := io.Pipe()
	session := &muxSession{id: id, in: in, out: out}
	m.sessions[id] = session
	m.wg.Add(1)
	go func() {
		defer m.wg.Done()
		rw := struct {
			io.Reader
			io.Writer
		}{out, &sessionWriter{mux: m, id: id}}
		_ = m.handle(ctx, rw)
		// The server is done with this connection: tell the helper so the client sees EOF instead
		// of a hang — unless the whole channel is going down, when the helper is gone or about to
		// be and a write it will never read must not hold the teardown.
		m.mu.Lock()
		closing := m.closing
		m.mu.Unlock()
		if !closing {
			_ = m.send(muxFrame{Conn: id, Event: "close"})
		}
		m.close(id)
	}()
}

func (m *mux) close(id int) {
	m.mu.Lock()
	session := m.sessions[id]
	delete(m.sessions, id)
	m.mu.Unlock()
	if session != nil {
		_ = session.in.Close()
	}
}

func (m *mux) closeAll() {
	m.mu.Lock()
	m.closing = true
	ids := make([]int, 0, len(m.sessions))
	for id := range m.sessions {
		ids = append(ids, id)
	}
	m.mu.Unlock()
	for _, id := range ids {
		m.close(id)
	}
	m.wg.Wait()
}

// sessionWriter tags a server reply with its connection and hands it to the helper. The server
// writes exactly one newline-terminated frame per call.
type sessionWriter struct {
	mux *mux
	id  int
}

func (w *sessionWriter) Write(p []byte) (int, error) {
	line := string(p)
	if n := len(line); n > 0 && line[n-1] == '\n' {
		line = line[:n-1]
	}
	if err := w.mux.send(muxFrame{Conn: w.id, Data: &line}); err != nil {
		return 0, err
	}
	return len(p), nil
}

File: internal/taskchannel/mux.js
'use strict';
// coop task-channel multiplexer — runs in the coop-owned HELPER container (never in the box).
//
// It listens on one unix socket that every MCP client in the box connects to through
// `socat STDIO UNIX-CONNECT:`, and carries all of those connections over its own stdio, which
// is coop's end of the channel. Frames are newline-delimited JSON, one per line, both ways:
//   helper -> coop   {"e":"ready"}                   the socket is bound and accepting
//                    {"c":N,"e":"open"}              client connection N appeared
//                    {"c":N,"d":"<one MCP line>"}    a line the client sent (no trailing newline)
//                    {"c":N,"e":"close"}             client connection N went away
//   coop -> helper   {"c":N,"d":"<one MCP line>"}    a line to deliver to client N
//                    {"c":N,"e":"close"}             hang up on client N
// EOF on stdin (coop is gone) closes every client and exits. The host side is mux.go.
const fs = require('fs');
const net = require('net');
const readline = require('readline');

const socketPath = process.argv[2];
if (!socketPath) {
  process.stderr.write('usage: mux.js <socket path>\n');
  process.exit(2);
}
// One client line may not exceed the same bound the host puts on a frame (4 MiB); a client that
// sends more without a newline is disconnected rather than buffered without limit.
const maxLine = 4 * 1024 * 1024;
const clients = new Map();
let nextID = 1;

function emit(frame) {
  process.stdout.write(JSON.stringify(frame) + '\n');
}

// allowHalfOpen: a client that shuts its write side (a `socat` whose stdin reached EOF after
// sending its requests) must still receive the replies coop is about to write. Without it node
// auto-closes the socket on read-EOF and the pending reply is lost. coop closes the socket itself
// when its session ends.
const server = net.createServer({ allowHalfOpen: true }, (socket) => {
  const id = nextID++;
  clients.set(id, socket);
  emit({ c: id, e: 'open' });
  socket.setEncoding('utf8');
  let pending = '';
  socket.on('data', (chunk) => {
    pending += chunk;
    let at;
    while ((at = pending.indexOf('\n')) >= 0) {
      emit({ c: id, d: pending.slice(0, at) });
      pending = pending.slice(at + 1);
    }
    if (pending.length > maxLine) socket.destroy();
  });
  socket.on('error', () => {});
  socket.on('close', () => {
    if (clients.delete(id)) emit({ c: id, e: 'close' });
  });
});
server.on('error', (err) => {
  process.stderr.write('coop task channel: ' + err.message + '\n');
  process.exit(1);
});
try { fs.unlinkSync(socketPath); } catch (e) { /* a fresh volume has no socket yet */ }
server.listen(socketPath, () => {
  // The box connects as the image's unprivileged user; the socket's inode must let it.
  fs.chmodSync(socketPath, 0o666);
  emit({ e: 'ready' });
});

const lines = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
lines.on('line', (line) => {
  let frame;
  try { frame = JSON.parse(line); } catch (e) { return; }
  const socket = clients.get(frame.c);
  if (!socket) return;
  if (frame.e === 'close') { socket.end(); return; }
  if (typeof frame.d === 'string') socket.write(frame.d + '\n');
});
lines.on('close', () => {
  for (const socket of clients.values()) socket.destroy();
  server.close(() => process.exit(0));
  setTimeout(() => process.exit(0), 1000).unref();
});
